activestorage/active.py

```diff
@@ -138,41 +125,55 @@ def __init__(
         :param storage_options: s3fs.S3FileSystem options
         :param active_storage_url: Reductionist server URL
         """
-        # Assume NetCDF4 for now
-        self.uri = uri
-        if self.uri is None:
-            raise ValueError(f"Must use a valid file for uri. Got {uri}")
+        self.ds = None
+        input_variable = False
+        if dataset is None:
+            raise ValueError(f"Must use a valid file or variable string for dataset. Got {dataset}")
+        if isinstance(dataset, Path) and not dataset.exists():
+            raise ValueError(f"Path to input file {dataset} does not exist.")
+        if not isinstance(dataset, Path) and not isinstance(dataset, str):
+            print(f"Treating input {dataset} as variable object.")
+            if not type(dataset) is pyfive.high_level.Dataset:
+                raise TypeError(f"Variable object dataset can only be pyfive.high_level.Dataset. Got {dataset}")
+            input_variable = True
+            self.ds = dataset
+        self.uri = dataset
+
 
         # still allow for a passable storage_type
         # for special cases eg "special-POSIX" ie DDN
         if not storage_type and storage_options is not None:
-            storage_type = urllib.parse.urlparse(uri).scheme
+            storage_type = urllib.parse.urlparse(dataset).scheme
         self.storage_type = storage_type
 
+        # set correct filename attr
+        if input_variable and not self.storage_type:
+            self.filename = self.ds
+        elif input_variable and self.storage_type == "s3":
+            self.filename = self.ds.id._filename
+
         # get storage_options
         self.storage_options = storage_options
         self.active_storage_url = active_storage_url
 
         # basic check on file
-        if not os.path.isfile(self.uri) and not self.storage_type:
-            raise ValueError(f"Must use existing file for uri. {self.uri} not found")
+        if not input_variable:
+            if not os.path.isfile(self.uri) and not self.storage_type:
+                raise ValueError(f"Must use existing file for uri. {self.uri} not found")
 
         self.ncvar = ncvar
-        if self.ncvar is None:
+        if self.ncvar is None and not input_variable:
             raise ValueError("Must set a netCDF variable name to slice")
 
         self._version = 1
         self._components = False
         self._method = None
         self._max_threads = max_threads
         self.missing = None
-        self.ds = None
-        self.metric_data = {}
         self.data_read = 0
 
-    @_metricise
     def __load_nc_file(self):
-        """ Get the netcdf file and it's b-tree"""
+        """ Get the netcdf file and its b-tree"""
         ncvar = self.ncvar
         # in all cases we need an open netcdf file to get at attributes
```

Comment:
I am not sure it's an open netcdf file any more is it?

Reply:
done in 61cf5dd31888519f78c07bbe4e948fb074c0310b